patches/server/0842-WIP-Brain-API.patch

```diff
@@ -0,0 +1,1020 @@
++
++    @SuppressWarnings({"unchecked", "rawtypes"})
++    @Override
++    public <T extends LivingEntity> void addSensor(@NotNull LivingEntity livingEntity, @NotNull SensorKey key, Sensor<? super T> sensor) {
++        Brain<?> brain = getBrain(livingEntity);
++
++        // See brain constructor
++        net.minecraft.world.entity.ai.sensing.Sensor mcSensor;
++        if (sensor instanceof PaperVanillaSensor vanillaSensor) { // Don't surround vanilla sensors
++            mcSensor = vanillaSensor.getHandle();
++        } else {
++            mcSensor = new PaperCustomSensor(sensor);
++        }
++
++        SensorType ignoreGenericType = PaperBrainUtil.getHandle(key);
++        // Register sensor
++        brain.sensors.put(ignoreGenericType, mcSensor);
++
++        // Register any required memories
++        Set<MemoryModuleType<?>> thanksIntellij = mcSensor.requires(); // Intellij wants to have this set iterate objects ???????????????????????????
++        for (MemoryModuleType<?> memoryType : thanksIntellij) {
++            brain.memories.putIfAbsent(memoryType, Optional.empty());
++        }
++    }
++
++    @Override
++    public void clearSensors(@NotNull LivingEntity livingEntity) {
++        Brain<?> brain = getBrain(livingEntity);
++        brain.sensors.clear();
++    }
++
++    @Override
++    public void removeSensor(@NotNull LivingEntity livingEntity, @NotNull SensorKey key) {
++        Brain<?> brain = getBrain(livingEntity);
++        brain.sensors.remove(PaperBrainUtil.getHandle(key));
++    }
++
++    @Nullable
++    @Override
++    public Sensor<? extends LivingEntity> getSensor(@NotNull LivingEntity livingEntity, @NotNull SensorKey key) {
++        Brain<?> brain = getBrain(livingEntity);
++        return brain.sensors.get(PaperBrainUtil.getHandle(key)).getPaperSensor();
++    }
++
++    @Override
++    public @NotNull Collection<Sensor<? extends LivingEntity>> getSensors(@NotNull LivingEntity livingEntity) {
++        Brain<?> brain = getBrain(livingEntity);
++
++        List<Sensor<? extends org.bukkit.entity.LivingEntity>> sensors = new ArrayList<>(brain.sensors.size());
++        for (net.minecraft.world.entity.ai.sensing.Sensor<?> sensor : brain.sensors.values()) {
++            sensors.add(sensor.getPaperSensor());
++        }
++
++        return sensors;
++    }
++
++    @Override
++    public void registerMemory(@NotNull LivingEntity livingEntity, @NotNull MemoryKey<?> memoryKey) {
++        Brain<?> brain = getBrain(livingEntity);
++        brain.memories.put(PaperBrainUtil.getHandle(memoryKey), Optional.empty());
++    }
++
++    @SuppressWarnings("unchecked")
++    @Override
++    public <V> void setMemory(@NotNull LivingEntity livingEntity, @NotNull MemoryKey<V> memoryKey, @Nullable V value, long expireIn) {
++        Brain<?> brain = getBrain(livingEntity);
++
++        if (memoryKey instanceof PaperMemory<V>) {
++            brain.setMemoryWithExpiry(PaperBrainUtil.getHandle(memoryKey), value, expireIn);
```

Comment:
Hmm should I make a new expiry method instead that is not null then?

Reply:
Redis has a concept we could follow.

You can 1. set a non-null value and 2. set its expiry. When 0 it will directly expire. If possible I would not use a long in API instead a duration to make conversion easier and calculation a non issue.
Defining two constant durations is recommended as well.
Like EXPIRE_NOW (0) and REMOVE_EXPIRY (-1).